### PR TITLE
[NeoMathEngine] Refactored MemoryEngine

### PR DESCRIPTION
**Pull request by favorart**
Added the class `CMemoryEngineMixin`, which incapsulates all of memory-pools and stack-allocators work for all `IMathEngine` derived classes.

### CODE REVIEW

File: NeoMathEngine/include/NeoMathEngine/NeoMathEngine.h [hunk 1]
```diff
@@ -1216,9 +1221,11 @@ class NEOMATHENGINE_API IMathEngine : public IDnnEngine {
 
 	// Typed data exchange
 	template<class T>
-	void DataExchangeTyped( const CTypedMemoryHandle<T>& result, const T* source, size_t size ) { DataExchangeRaw( result, source, size * sizeof(T) ); }
+	void DataExchangeTyped( const CTypedMemoryHandle<T>& result, const T* source, size_t size )
+	{ DataExchangeRaw( result, source, size * sizeof(T) ); }
```
Comment (AndrewAndrianov) [inline NeoMathEngine/include/NeoMathEngine/NeoMathEngine.h]: +tab

File: NeoMathEngine/src/MemoryEngine.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+#include <mutex>
+
+namespace NeoML {
+
+class CDeviceStackAllocator;
+class CHostStackAllocator;
+class CMemoryPool;
+
+// Memory management engine base class
+class IMemoryEngine : public IMathEngine {
```
Comment (AndrewAndrianov) [inline NeoMathEngine/src/MemoryEngine.h]: Not an interface - it is base class.
I propose name `CMemoryEngineMixin`

File: NeoMathEngine/src/MemoryEngine.cpp [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+	CleanUpSpecial();
+	MemoryPool->CleanUp();
+}
+
+void* IMemoryEngine::GetBuffer( const CMemoryHandle& handle, size_t pos, size_t size, bool exchange )
+{
+	ASSERT_EXPR( HostStackAllocator != nullptr );
+	ASSERT_EXPR( handle.GetMathEngine() == this );
+
+	size_t realSize = size + 16;
```
Comment (AndrewAndrianov) [inline NeoMathEngine/src/MemoryEngine.cpp]: magic constant 16

**Review by AndrewAndrianov — CHANGES_REQUESTED** [review]

File: NeoMathEngine/src/MemoryEngine.h (hunk 2 above)
Comment (favorart) [reply NeoMathEngine/src/MemoryEngine.h]: fixed

File: NeoMathEngine/src/MemoryEngine.cpp (hunk 3 above)
Comment (favorart) [reply NeoMathEngine/src/MemoryEngine.cpp]: fixed

**Review by AndrewAndrianov — APPROVED** [review]

**Pull request merged**